django/cantusdb_project/main_app/tests/test_functions.py

```diff
@@ -118,3 +260,156 @@ def test_generate_incipit(self):
+                "NotACantusID"
+            )
+        with self.subTest(test="Ensure None returned in case of nonexistent Cantus ID"):
+            self.assertIsNone(suggested_chants_nonexistent_cantus_id)
+
+        with patch("requests.get", mock_requests_get):
+            suggested_chants_rare_cantus_id = get_suggested_chants(cantus_id="a07763")
+        with self.subTest(
+            test="Ensure None is returned in case of Cantus ID without suggestions"
+        ):
+            self.assertIsNone(suggested_chants_rare_cantus_id)
+
+    def test_get_json_from_ci_api(self):
+        with patch("requests.get", mock_requests_get):
+            json_nextchants_response = get_json_from_ci_api(
+                path="/json-nextchants/001010"
+            )
+        with self.subTest(
+            test="Ensure properly handles /nextchants/<cantus_id> endpoint"
+        ):
+            self.assertIsInstance(json_nextchants_response, list)
+            first_nextchant = json_nextchants_response[0]
+            self.assertIsInstance(first_nextchant, dict)
+
+        with patch("requests.get", mock_requests_get):
+            json_cid_response = get_json_from_ci_api(path="/json-cid/008349")
+        observed_json_cid_keys = json_cid_response.keys()
+        expected_json_cid_keys = ("info", "chants")
+        with self.subTest(
+            test="Ensure properly handles /json-cid/<cantus_id> endpoint"
+        ):
+            self.assertIsInstance(json_cid_response, dict)
+            for key in expected_json_cid_keys:
+                self.assertIn(key, observed_json_cid_keys)
+
+        with patch("requests.get", mock_requests_get):
+            response_short_timeout = get_json_from_ci_api(
+                path="/some/path", timeout=0.0001
+            )
+        with self.subTest(test="Ensure returns None when requests.get times out"):
+            self.assertIsNone(response_short_timeout)
+
+        with patch("requests.get", mock_requests_get):
+            response_nonexistent_cantus_id = get_json_from_ci_api(
+                path="/json-cid/notACantusID"
+            )
+        with self.subTest(
+            test="Ensure returns None when response status code is not 200"
+        ):
+            self.assertIsNone(response_nonexistent_cantus_id)
+
+        # I can't figure out how to get assertRaises to work - even when the assertion should fail, it doesn't.
+        # It's not of vital importance that this argument check work correctly, I think, so I'm giving up in
+        # an effort not to get bogged down.
+        # Leaving a couple of commented-out attempts here with things I've tried, for the hopeful benefit of some
+        # future, cleverer developer
+        # - Jacob dGM, April 2024
+
+        # with self.subTest(test="Ensure raises ValueError when path improperly formatted, attempt 1"):
+        #     self.assertRaises(ValueError, get_json_from_ci_api, "/for/some/reason/this/passes/even/with/a/leading/slash")
```

Comment:
As I read it, your function `get_json_from_ci_api` would raise `ValueError` if there *isn't* a leading slash.

Reply:
Right. So considering there _is_ a leading slash, and no `ValueError` is raised, we expect `AssertRaises` to fail. And yet it doesn't, for reasons I don't understand.

As far as I can figure, the most important part of a test suite is not that all the tests pass, but that the tests fail when they're supposed to. Before writing `self.assertRaises(ValueError, get_json_from_ci_api, "path/without/leading/slash")`, which passes, I want to be sure that `self.assertRaises(ValueError, get_json_from_ci_api, "/path/with/leading/slash")` fails.

Reply:
Ok, so this test does fail for me:

![image](https://github.com/DDMAL/CantusDB/assets/11023634/6cb64be5-ec04-46bf-a525-55cc9f80ad34)

Reply:
It seems to me that if what I'm getting is the desired outcome, then this would be a place for an `expectedFailure` decorator?

Reply:
> I've also been having problems with one `ChantModelTest.test_get_concordances` - if I run `python manage.py test main_app.tests` as a whole, it errors, but if I run `python manage.py test main_app.tests.test_models.ChantModelTest`, it passes.

You should definitely make an issue about this...this is bad!

> Did you try the `self.AssertRaises(several_arguments)` version or the `with self.AssertRaises(ValueError)` version? Should I just go ahead and uncomment that one and delete the other?

The latter.

Reply:
oh! I think I figured it out.

This assertion was behaving weirdly because `"/path/with/leading/slash"` or whatever was not in the list of paths that `mock_requests_get` was set up to handle. Now, `mock_requests_get` raises a `NotImplementedError` when a non-standard URL path is passed, but previously, it raised a - you guessed it - `ValueError`.